Adds support for Rulesets having intrinsic template requirements

ZephyrTFA opened the pull request:
Title

## About The Pull Request

Ensures that we load templates for a ruleset before we attempt to place or cache characters for that ruleset
Also makes wizard and abductor async load their template to improve (apparent) loading times for them
## Why It's Good For The Game

This is the only thing left that I can think of that would cause antags like nukies and abductors to spawn in wrong
## Changelog
This should not be player facing

ZephyrTFA commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -198,11 +199,13 @@
+	if (required_candidates <= living_players.len)
+		return TRUE
+
+	log_dynamic("FAIL: [src], a from_ghosts ruleset, did not have enough dead candidates: [required_candidates] needed, [dead_count] found")
```
```suggestion
	log_dynamic("FAIL: [src], a from_living ruleset, did not have enough candidates: [required_candidates] needed, [living_players.len] found")
```

Mothblocks commented on code/__DEFINES/dynamic.dm:
```diff
@@ -7,6 +7,9 @@
 #define LONE_RULESET (1 << 2)
 
+/// This ruleset has lazy loaded templates that must be loaded if the ruleset is selected
+#define RULESET_USES_LAZY_TEMPLATES (1 << 3)
```
This flag doesn't feel necessary

ZephyrTFA replied:
I can have it just check for the list existing?

Mothblocks replied:
You don't even haev to do that, looping over null does nothing

Mothblocks commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -170,8 +173,12 @@
 /// Remember that on roundstart no one knows what their job is at this point.
 /// IMPORTANT: If ready() returns TRUE, that means pre_execute() or execute() should never fail!
-/datum/dynamic_ruleset/proc/ready(forced = 0)
-	return check_candidates()
```
Hmmmm are we sure that forcing everything to call this doesn't change behavior at all? I am fairly sure some of the calls without were intentional

ZephyrTFA replied:
how about just loading the templates before it calls ready?

Mothblocks replied:
Yeah

ZephyrTFA committed:
ensure only one template can load at a time

ZephyrTFA committed:
make shuttle loading async, but block on final dock

ZephyrTFA committed:
ruleset lazy templates

ZephyrTFA committed:
remove no longer needed lazy load calls
auto make some of them async on datum creation to improve (apparent) loading times

ZephyrTFA committed:
wrong var

Mothblocks commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -853,6 +853,16 @@ GLOBAL_LIST_EMPTY(the_station_areas)
+
+	lazy_loading = TRUE
+	. = _lazy_load_template(template_key, force)
+	lazy_loading = FALSE
```
```suggestion
	lazy_loading = FALSE
	return .
```

Mothblocks commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -273,10 +273,12 @@
-	SSmapping.lazy_load_template(LAZY_TEMPLATE_KEY_WIZARDDEN)
-	if(GLOB.wizardstart.len == 0)
+/datum/dynamic_ruleset/roundstart/wizard/ready(forced = FALSE)
+	if(!check_candidates())
```
I'm still not sure we should be changing how the rulesets act in this PR, old Dynamic code is pretty clunky and I wouldn't be surprised if this broke something

ZephyrTFA replied:
Just made it copy the other ruleset logic

Mothblocks reviewed (changes requested):
Looks good other than this

ZephyrTFA committed:
Update code/controllers/subsystem/mapping.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>
